samples/net/sockets/echo_client/src/echo-client.c

```diff
@@ -182,3 +172,71 @@ void main(void)
 		stop_tcp();
 	}
 }
+
+static void do_service(k_work_handler_t handler)
+{
+	static struct k_work work;
+
+	k_work_init(&work, handler);
+
+	k_work_submit(&work);
+}
+
+static void event_handler(struct net_mgmt_event_callback *cb,
+			  u32_t mgmt_event, struct net_if *iface)
+{
+	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
+		LOG_INF("Network connected");
+
+		do_service(start_udp_and_tcp);
```

Comment:
As I read this, we're moving the main() loop to start_udp_and_tcp().  This means the main thread of the sample is now wasted (ends up in a forever sleep).  And the primary processing of the sample ends up on the system work queue.   As an example sample which will be copy/pasted (potentially) -- this feels wrong to me.

In order to preserve the original main thread flow of the sample, it seems like we should be enter the regular while loop flow, but have thread-safe primitives in the event handler which pause and restart the main thread.

@andyross might have an opinion on this as well.

If this were a one off sample this might be fine.  But, I think we're establishing the right way to do this in samples.